type-check.ts

```diff
@@ -63,19 +73,81 @@ export type TypeError = {
   message: string
 }
 
-export function equalType(t1: Type, t2: Type) {
+// combine the elements of two arrays into an array of tuples.
+// DANGER: behaves weirdly with arrays of different length.
+function zip<A, B>(l1: Array<A>, l2: Array<B>) : Array<[A, B]> {
+  return l1.map((el, i) => [el, l2[i]]); 
+}
+
+export function equalType(env: GlobalTypeEnv, t1: Type, t2: Type) {
+  // If either of t1 or t2 are type-variables, then the actual type
+  // that they are currently instantiated to needs to be fetched from
+  // the environment before comparing the types.
+  if(t1.tag === "typevar") {
+    t1 = env.typevars.get(t1.name)[2];
+  }
+
+  if(t2.tag === "typevar") {
+    t2 = env.typevars.get(t2.name)[2];
+  }
+  
   return (
     t1 === t2 ||
-    (t1.tag === "class" && t2.tag === "class" && t1.name === t2.name)
+    (t1.tag === "class" && t2.tag === "class" && t1.name === t2.name && equalTypeParams(env, t1.params, t2.params))
   );
 }
 
-export function isNoneOrClass(t: Type) {
-  return t.tag === "none" || t.tag === "class";
+// Check if a list of type-parameters are equal.
+export function equalTypeParams(env: GlobalTypeEnv, params1: Type[], params2: Type[]) : boolean {
+  if(params1.length !== params2.length) {
+    return false;
+  }
+
+  return zip(params1, params2).reduce((isEqual, [p1, p2]) => {
+    return isEqual && equalType(env, p1, p2);
+  }, true);
+}
+
+export function isNoneOrClass(env: GlobalTypeEnv, t: Type) {
+  if(t.tag === "typevar") {
+    // If t is a type-variables, then the actual type
+    // that it is currently instantiated to needs to be fetched from
+    // the environment.
+    t = env.typevars.get(t.name)[2];
+  }
+
+  // TODO: decide if something with type-parameters is considered a class
+  return t.tag === "none" || (t.tag === "class" && t.name !== UNCONSTRAINED);
+}
+
+export function getClassName(env: GlobalTypeEnv, t: Type): string {
+  if(t.tag === "typevar") {
+    // If t is a type-variables, then the actual type
+    // that it is currently instantiated to needs to be fetched from
+    // the environment.
+    t = env.typevars.get(t.name)[2];
+  }
+
+  if(t.tag !== "class" || t.name === UNCONSTRAINED) {
+    return undefined;
```

Comment:
Oof I don't like using `undefined` as a meaningful value here. Very confusing. What does this mean?

Reply:
This means that we are trying to get the class name of something that is not an object. This is equivalent to the earlier explicit check for "lookup"/"method-call" where we were checking if the lhs expression has the tag class, only it gets hidden behind this function that returns the class name. 

I think I initially implemented this as a function that returned Type but this needed me to do a redundant comparison on the outside to satisfy typescript for some reason... As I write this I feel like it should be possible to clean it up in that way without requiring any redundant comparison. Let me try to do that.